File: merge_sort.js
let arr = [2,5,3,6,4,2,6,3,8,5]


function merge(bottom, top) {
    console.log('merge', bottom, top)
    let smallest = 0
    if (top.length === 0 ) {
        return bottom
    } 
    if (bottom.length === 0) {
        return top
    }

    smallest = bottom[0] <= top[0] ? bottom.shift() : top.shift()

    let merged = merge(bottom, top)
    merged.unshift(smallest)
    console.log(merged)
    return merged

}

function mergeSort(arr) {
    if (arr.length <= 1) {
        console.log('base', arr)
        return arr
    }
    let middle = Math.ceil(arr.length / 2)
    let topHalf = arr.slice(middle,arr.length);
    let bottomHalf = arr.slice(0,middle);
    console.log(topHalf, bottomHalf)
    let sortedTopHalf = mergeSort(topHalf)
    let sortedBottomHalf = mergeSort(bottomHalf)
    
    return merge(sortedBottomHalf, sortedTopHalf)
}
    
console.log(mergeSort(arr));

File: merge_sort.py
# Merge  Sort
from random import shuffle

arr = list(range(25))
shuffle(arr)


def merge(left, right):
    print('start of merge', left, right)
    smallest = 0
    # base cases
    if len(right) == 0:
        return left
    if len(left) == 0:
        return right
    # compare and find smallest number and remove it
    if left[0] <= right[0]:
        smallest = left.pop(0)
    else:
        smallest = right.pop(0)

    # keep merging until either list is empty
    merged = merge(left, right)
    merged.insert(0, smallest)
    print('merged', merged)
    return merged


def merge_sort(l):
    # base case
    # a list of one or fewer is considered sorted
    if len(l) <= 1:
        return l

    mid = round(len(l) / 2)
    # divide list around midpoint
    left_list = l[:mid]
    right_list = l[mid:]
    # print(left_list, right_list)
    sorted_left_list = merge_sort(left_list)
    sorted_right_list = merge_sort(right_list)
    # print(sorted_left_list, sorted_right_list)

    return merge(sorted_left_list, sorted_right_list)


merge_sort(arr)
